File: engine/meicho/worlds.py
"""相手の手札としてありうる世界の**数え上げと列挙**（文献計画 便 C 段 C-3・D-077 追記 3）。

## なぜここに移したか

W（区別できる相手の手札の数）の数え方は、もともと診断の道具
`experiments/diag_pimc.py` の中にあった（便 M）。段 C-3（II-9 終盤の全列挙）では
**探索の本体**が同じ数え方を使う。写して 2 か所に置くと、片方だけ直したときに
「診断が言う W」と「探索が使う W」が黙ってずれる（D-075 の `TracingPlanner` の教訓）。
そこで**数える関数はここ 1 か所に置き、`diag_pimc.py` はここを呼ぶ**。

数字が変わっていないことは `tests/test_lit_c.py::test_worlds_module_matches_diag_pimc` が
便 M の記録（`results/lit/m5m1_pimc.jsonl`）と突き合わせて固定する。

## 何を数えるか

相手の手札は**多重集合**である（同じ札が 2 枚あってもその 2 枚は区別しない）。
公開情報から作った未公開の候補 `pool` と、手札の枚数 `n_hand`、
そして「必ず含まれると分かっている札」`known`（スキャンで見て今も手札にある札・D-023）から、

- **W** … ありうる区別できる手札が何通りあるか
- **H_w** … 多重度で重みづけた分布のエントロピー（bit）
- **列挙** … その手札を全部並べる（重み＝その型になる**物理的な配り方**の数）

を出す。重みは Π_i C(c_i, a_i)——「その種類の札が c_i 枚あるうち a_i 枚を持つ組み合わせ」の積である。
配りが一様なら、手札 h の確率はこの重みに比例する。

**列挙は W が小さいときにしか使えない。** 段 C-3 は W ≤ 64 の決定に限って使う
（便 M の実測で決定の 14.3%・終盤に集中する）。
"""
from __future__ import annotations

import math
from collections import Counter
from math import comb


def remove_multiset(pool: list, take) -> list:
    """多重集合の引き算。`pool` から `take` の各カードを 1 枚ずつ取り除いた残り。

    並びは `pool` のまま（取り除いたものが抜けるだけ）。
    """
    left = Counter(take)
    out = []
    for cid in pool:
        if left.get(cid, 0) > 0:
            left[cid] -= 1
        else:
            out.append(cid)
    return out


def world_counts(pool, n_hand: int, known=()) -> dict:
    """相手の手札としてありうる**区別できる多重集合**の数 W と、重みつきエントロピー H_w。

    数え上げは種類ごとの生成関数 ∏(1 + x + … + x^{c_i}) の x^k の係数で、
    同じ DP で Σw と Σ w·ln w を持ち回ってエントロピーも一度に出す。

    **この関数の中身は便 M（`diag_pimc.py`）から 1 行も変えずに移したものである。**
    """
    pool = list(pool)
    known = list(known)
    left = remove_multiset(pool, known)
    k = n_hand - len(known)
    if k < 0 or k > len(left):
        # 情報が食い違う（起きないはずだが、起きたら黙って 0 にしない）
        return {"W": 0, "H_w": None, "k": k, "pool_size": len(pool),
                "known_n": len(known), "inconsistent": True}
    S = [0.0] * (k + 1)          # Σ 重み
    T = [0.0] * (k + 1)          # Σ 重み·ln(重み)
    W = [0] * (k + 1)            # 区別できる多重集合の数
    S[0], W[0] = 1.0, 1
    for c in sorted(Counter(left).values()):
        nS, nT, nW = [0.0] * (k + 1), [0.0] * (k + 1), [0] * (k + 1)
        for j in range(k + 1):
            if W[j] == 0:
                continue
            for a in range(0, min(c, k - j) + 1):
                v = float(comb(c, a))
                nS[j + a] += S[j] * v
                nT[j + a] += v * T[j] + S[j] * v * math.log(v)
                nW[j + a] += W[j]
        S, T, W = nS, nT, nW
    tot = S[k]
    h = None if tot <= 0 else (math.log(tot) - T[k] / tot) / math.log(2.0)
    if h is not None and abs(h) < 1e-12:
        h = 0.0
    return {"W": W[k], "H_w": h, "k": k, "pool_size": len(pool),
            "known_n": len(known), "inconsistent": False}


def enumerate_hands(pool, n_hand: int, known=(), limit: int = None) -> list:
    """ありうる相手の手札を**全部**並べる（段 C-3・II-9）。

    返すのは `[(手札, 重み), ...]`。手札は `known` を含む**並べ替え済み**のリストで、
    重みは「その型になる物理的な配り方の数」（`world_counts` の重みと同じ定義）。
    並びは **重みの大きい順、同点は手札の並び順**——`limit` で上から切っても
    同じシードなら同じ集合になるように、**完全に決定的**にしてある。

    `limit` を渡すと上位 `limit` 本だけ返す（§0.3 (vi) の `endgame_eval`）。
    W ≤ limit なら**取りこぼしが無い＝厳密**である。

    数え上げ（`world_counts`）と同じ規則で作ってあることは
    `tests/test_lit_c.py::test_endgame_enumeration_is_exhaustive` が固定する
    （列挙した本数が W と一致し、重みの合計が数え上げの Σ 重みと一致する）。
    """
    pool = list(pool)
    known = list(known)
    left = remove_multiset(pool, known)
    k = n_hand - len(known)
    if k < 0 or k > len(left):
        return []
    kinds = sorted(Counter(left).items())          # (card_id, 枚数) を id 順に
    out: list = []

    def walk(i: int, rest: int, taken: list, weight: float) -> None:
        if rest == 0:
            hand = sorted(known + [c for c, a in taken for _ in range(a)])
            out.append((hand, weight))
            return
        if i >= len(kinds):
            return
        cid, cnt = kinds[i]
        # 残りの札で足りるかを先に見る（枝を無駄に伸ばさない）
        if sum(c for _, c in kinds[i:]) < rest:
            return
        for a in range(min(cnt, rest), -1, -1):
            walk(i + 1, rest - a, taken + ([(cid, a)] if a else []),
                 weight * comb(cnt, a))

    walk(0, k, [], 1.0)
    # 重みの大きい順・同点は手札の並び順（**完全に決定的**）
    out.sort(key=lambda t: (-t[1], t[0]))
    return out if limit is None else out[:max(0, int(limit))]


def d_from_w(w: int, w0: int) -> float:
    """D_t = 1 − log W_t / log W_0。W_0 = 1（最初から確定）なら 1 と定義する。"""
    if not w0 or w0 <= 1:
        return 1.0
    if w <= 0:
        return 1.0
    return 1.0 - math.log(w) / math.log(w0)

File: engine/rust/src/worlds.rs
//! 相手の手札としてありうる世界の**数え上げと列挙**（文献計画 便 C 段 C-3・II-9）。
//!
//! **真実源は Python の `meicho/worlds.py`。ここはその写しである。**
//! 並び（`sorted`）はカード ID の文字列順であり、Rust では `db.action_rank` で表す
//! （`agents.rs::Greedy::sort_by_id` と同じ規約）。数え方・重みの定義・切り方まで
//! Python と 1 対 1 に保つこと。ずれると毎手一致の検査が「同じ手だが別の理由」で崩れる。

use crate::cards::CardDb;

/// 多重集合の引き算（`pool` から `take` を 1 枚ずつ）。並びは `pool` のまま。
pub fn remove_multiset(pool: &[u16], take: &[u16]) -> Vec<u16> {
    let mut left: Vec<u16> = take.to_vec();
    let mut out: Vec<u16> = Vec::with_capacity(pool.len());
    for &cid in pool {
        match left.iter().position(|&x| x == cid) {
            Some(ix) => {
                left.remove(ix);
            }
            None => out.push(cid),
        }
    }
    out
}

/// 二項係数 C(n, k)（小さい値しか使わないので f64 で十分）。
fn comb(n: usize, k: usize) -> f64 {
    if k > n {
        return 0.0;
    }
    let k = k.min(n - k);
    let mut v = 1.0f64;
    for i in 0..k {
        v = v * ((n - i) as f64) / ((i + 1) as f64);
    }
    v.round()
}

/// 種類ごとの枚数（カード ID 順）。`sorted(Counter(left).items())` の写し。
fn kinds(db: &CardDb, left: &[u16]) -> Vec<(u16, usize)> {
    let mut ks: Vec<(u16, usize)> = Vec::new();
    for &c in left {
        match ks.iter_mut().find(|(x, _)| *x == c) {
            Some(e) => e.1 += 1,
            None => ks.push((c, 1)),
        }
    }
    ks.sort_by_key(|(c, _)| db.action_rank[*c as usize]);
    ks
}

/// ありうる**区別できる手札**の数 W。情報が食い違うときは `None`。
///
/// Python の `world_counts(...)["W"]` と同じ値。エントロピー H_w は探索が使わないので
/// ここでは持たない（診断側は Python の `worlds.py` を呼ぶ）。
pub fn world_count_w(db: &CardDb, pool: &[u16], n_hand: usize, known: &[u16]) -> Option<u64> {
    let left = remove_multiset(pool, known);
    if n_hand < known.len() {
        return None;
    }
    let k = n_hand - known.len();
    if k > left.len() {
        return None;
    }
    let mut w: Vec<u64> = vec![0; k + 1];
    w[0] = 1;
    let mut counts: Vec<usize> = kinds(db, &left).iter().map(|(_, c)| *c).collect();
    counts.sort_unstable();
    for c in counts {
        let mut nw: Vec<u64> = vec![0; k + 1];
        for j in 0..=k {
            if w[j] == 0 {
                continue;
            }
            for a in 0..=c.min(k - j) {
                nw[j + a] = nw[j + a].saturating_add(w[j]);
            }
        }
        w = nw;
    }
    Some(w[k])
}

/// ありうる手札を**全部**並べる。返すのは `(手札, 重み)` で、手札はカード ID 順に並べた
/// `known` 込みのもの、重みは「その型になる物理的な配り方の数」。
/// 並びは**重みの大きい順・同点は手札の並び順**（切っても決定的）。
/// `limit` は 0 のとき「切らない」を意味する（Python の `limit=None`）。
pub fn enumerate_hands(db: &CardDb, pool: &[u16], n_hand: usize, known: &[u16],
                       limit: usize) -> Vec<(Vec<u16>, f64)> {
    let left = remove_multiset(pool, known);
    if n_hand < known.len() {
        return Vec::new();
    }
    let k = n_hand - known.len();
    if k > left.len() {
        return Vec::new();
    }
    let ks = kinds(db, &left);
    let mut known_sorted = known.to_vec();
    known_sorted.sort_by_key(|&c| db.action_rank[c as usize]);
    let mut out: Vec<(Vec<u16>, f64)> = Vec::new();
    let mut taken: Vec<(u16, usize)> = Vec::new();
    walk(db, &ks, 0, k, &mut taken, 1.0, &known_sorted, &mut out);
    // 重みの大きい順（同点はカード ID 順の手札の並び順）
    out.sort_by(|a, b| {
        b.1.partial_cmp(&a.1).unwrap().then_with(|| {
            let ra: Vec<u32> = a.0.iter().map(|&c| db.action_rank[c as usize] as u32).collect();
            let rb: Vec<u32> = b.0.iter().map(|&c| db.action_rank[c as usize] as u32).collect();
            ra.cmp(&rb)
        })
    });
    if limit > 0 && out.len() > limit {
        out.truncate(limit);
    }
    out
}

#[allow(clippy::too_many_arguments)]
fn walk(db: &CardDb, ks: &[(u16, usize)], i: usize, rest: usize,
        taken: &mut Vec<(u16, usize)>, weight: f64, known: &[u16],
        out: &mut Vec<(Vec<u16>, f64)>) {
    if rest == 0 {
        let mut hand: Vec<u16> = known.to_vec();
        for (cid, a) in taken.iter() {
            for _ in 0..*a {
                hand.push(*cid);
            }
        }
        hand.sort_by_key(|&c| db.action_rank[c as usize]);
        out.push((hand, weight));
        return;
    }
    if i >= ks.len() {
        return;
    }
    let (cid, cnt) = ks[i];
    // 残りの札で足りるかを先に見る（Python と同じ枝刈り）
    let avail: usize = ks[i..].iter().map(|(_, c)| *c).sum();
    if avail < rest {
        return;
    }
    for a in (0..=cnt.min(rest)).rev() {
        if a > 0 {
            taken.push((cid, a));
        }
        walk(db, ks, i + 1, rest - a, taken, weight * comb(cnt, a), known, out);
        if a > 0 {
            taken.pop();
        }
    }
}
